packages/regl-worldview/src/commands/GLText.js

```diff
@@ -0,0 +1,393 @@
+    }
+
+    if (gl_FragColor.a == 0.) {
+      discard;
+    }
+  }
+`;
+
+function makeTextCommand() {
+  // Keep the set of rendered characters around so we don't have to rebuild the font atlas too often.
+  const charSet = new Set();
+  const memoizedBuildAtlas = createMemoizedBuildAtlas();
+
+  const command = (regl: any) => {
+    const atlasTexture = regl.texture();
+    const drawText = regl({
+      depth: defaultDepth,
+      blend: defaultBlend,
+      primitive: "triangle strip",
+      vert,
+      frag,
+      uniforms: {
+        atlas: atlasTexture,
+        atlasSize: () => [atlasTexture.width, atlasTexture.height],
+        fontSize: FONT_SIZE,
+        cutoff: CUTOFF,
+        outlineCutoff: OUTLINE_CUTOFF,
+        srcHeight: FONT_SIZE + 2 * BUFFER,
+      },
+      instances: regl.prop("instances"),
+      count: 4,
+      attributes: {
+        position: [[0, 0], [0, -1], [1, 0], [1, -1]],
+        texCoord: [[0, 0], [0, 1], [1, 0], [1, 1]], // flipped
+        srcOffset: (ctx, props) => ({ buffer: props.srcOffsets, divisor: 1 }),
+        destOffset: (ctx, props) => ({ buffer: props.destOffsets, divisor: 1 }),
+        srcWidth: (ctx, props) => ({ buffer: props.srcWidths, divisor: 1 }),
+        scale: (ctx, props) => ({ buffer: props.scale, divisor: 1 }),
+
+        alignmentOffset: (ctx, props) => ({ buffer: props.alignmentOffset, divisor: 1 }),
+        billboard: (ctx, props) => ({ buffer: props.billboard, divisor: 1 }),
+
+        foregroundColor: (ctx, props) => ({ buffer: props.foregroundColor, divisor: 1 }),
+        outlineColor: (ctx, props) => ({ buffer: props.outlineColor, divisor: 1 }),
+        enableOutline: (ctx, props) => ({ buffer: props.enableOutline, divisor: 1 }),
+
+        posePosition: (ctx, props) => ({ buffer: props.posePosition, divisor: 1 }),
+        poseOrientation: (ctx, props) => ({ buffer: props.poseOrientation, divisor: 1 }),
+      },
+    });
+
+    return (props: $ReadOnlyArray<TextMarker & { billboard?: boolean }>) => {
+      const prevNumChars = charSet.size;
+      for (const { text } of props) {
+        for (const char of text) {
+          charSet.add(char);
+        }
+      }
+      const charsChanged = charSet.size !== prevNumChars;
+
+      const { textureData, textureWidth, textureHeight, charInfo } = memoizedBuildAtlas(
+        // only use a new set if the characters changed, since memoizeOne uses shallow equality
+        charsChanged ? new Set(charSet) : charSet
+      );
+
+      // re-upload texture only if characters were added
+      if (charsChanged) {
+        atlasTexture({
+          data: textureData,
+          width: textureWidth,
+          height: textureHeight,
+          format: "alpha",
+          wrap: "clamp",
+          mag: "linear",
+          min: "linear",
+        });
+      }
+
+      let totalInstances = 0;
+      const estimatedInstances = props.reduce((sum, marker) => sum + marker.text.length, 0);
```

Comment:
you could save a loop and do this when you go through each char on line 252

Reply:
point here is to know what size buffer to allocate up front

Reply:
yah - aren't you computing the entire character count of all characters in each text marker?  I was saying you're _also_ looping through each text marker around line 252.  You could keep the count up there & not have to loop again.  It's a very small optimization by removing a loop.

Reply:
oh gotcha! Should've checked the line number 🤣